tracecat/llm/ollama.py

```diff
@@ -0,0 +1,129 @@
+    MISTRAL_LARGE = "mistral-large:123b"
+    MIXTRAL = "mixtral:8x7b"
+    MIXTRAL_22B = "mixtral:8x22b"
+
+
+DEFAULT_OLLAMA_MODEL = OllamaModel.LLAMA32_1B
+
+
+def _get_ollama_client(api_url: str | None = None) -> ollama.AsyncClient:
+    return ollama.AsyncClient(host=api_url or config.OLLAMA__API_URL)
+
+
+async def preload_ollama_models(models: list[str]) -> list[dict[str, Any]]:
+    client = _get_ollama_client()
+    responses = []
+    # Download iteratively to avoid overwhelming
+    # the Ollama client and server
+    for model in models:
+        try:
+            await client.pull(model)
+        except Exception as e:
+            logger.warning(
+                "Failed to pull model",
+                model=model,
+                error=e,
+            )
+    return responses
+
+
+async def check_model_valid(model: str, api_url: str | None = None) -> bool:
+    """Check if a model is valid."""
+    client = _get_ollama_client(api_url)
+    try:
+        await client.pull(model)
+    except Exception:
+        return False
+    return True
+
+
+async def list_local_models(api_url: str | None = None) -> list[dict[str, Any]]:
+    """List all models available locally."""
+    client = _get_ollama_client(api_url)
+    models = await client.list()
+    return [model.model_dump() for model in models.models]
+
+
+async def list_local_model_names(api_url: str | None = None) -> list[str | None]:
+    """List all model names available locally."""
+    models = await list_local_models(api_url)
+    return [model["model"] for model in models]
+
+
+async def is_local_model(model: str, api_url: str | None = None) -> bool:
+    """Check if a model is available locally."""
+    return model in await list_local_model_names(api_url)
+
+
+async def async_ollama_call(
+    prompt: str,
+    *,
+    model: OllamaModel = DEFAULT_OLLAMA_MODEL,
+    memory: list[dict[str, Any]] | None = None,
+    system_prompt: str | None = None,
+    format: BaseModel | None = None,
+    api_url: str | None = None,
+) -> ChatResponse:
+    client = _get_ollama_client(api_url)
+
+    if not await is_local_model(model, api_url):
+        logger.warning(
+            "Local LLM model not found",
+            provider="ollama",
+            model=model,
+        )
+        raise ValueError(f"Local LLM model {model!r} not found")
+
+    if system_prompt:
+        messages = [
+            {"role": "system", "content": system_prompt},
+            {"role": "user", "content": prompt},
+        ]
+    elif memory:
+        messages = [
+            *memory,
+            {"role": "user", "content": prompt},
+        ]
+    else:
+        messages = [{"role": "user", "content": prompt}]
+
+    logger.debug(
+        "🧠 Calling LLM chat completion",
+        provider="ollama",
+        model=model,
+        prompt=prompt,
+    )
+
+    kwargs = {
+        "model": model,
+        "messages": messages,
+        "format": format.model_json_schema() if format else None,
```

Comment:
does passing `None` here have the same effect as omitting the kwarg?

Reply:
No idea. Used conditional instead to be safe.